br/pkg/restore/data.go

```diff
@@ -140,138 +149,118 @@ func getStoreAddress(allStores []*metapb.Store, storeId uint64) string {
 }
 
 // read all region meta from tikvs
-func (recovery *Recovery) ReadRegionMeta(ctx context.Context, totalTiKVs int, allStores []*metapb.Store) error {
+func (recovery *Recovery) ReadRegionMeta(ctx context.Context) error {
 	eg, ectx := errgroup.WithContext(ctx)
-	workers := utils.NewWorkerPool(uint(totalTiKVs), "Collect Region Meta")
+	totalStores := len(recovery.allStores)
+	workers := utils.NewWorkerPool(uint(mathutil.Min(totalStores, max_store_concurency)), "Collect Region Meta") // TODO: int overflow?
 
 	// TODO: optimize the ErroGroup when TiKV is panic
-	metaChan := make(chan RecoveryMeta, 1024)
+	metaChan := make(chan StoreMeta, 1024)
 	defer close(metaChan)
 
-	for i := 0; i < totalTiKVs; i++ {
+	for i := 0; i < totalStores; i++ {
 		i := i
-		storeId := allStores[i].GetId()
-		tikvClient, err := recovery.newTiKVRecoveryClient(ectx, allStores[i].GetAddress())
+		storeId := recovery.allStores[i].GetId()
+		storeAddr := recovery.allStores[i].GetAddress()
+
+		if err := ectx.Err(); err != nil {
+			return errors.Trace(err)
+		}
+
+		tikvClient, conn, err := recovery.newTiKVRecoveryClient(ectx, storeAddr)
 		if err != nil {
-			log.Error("create tikv client failied", zap.Uint64("storeID", storeId))
 			return errors.Trace(err)
 		}
-		log.Debug("read meta from tikv", zap.String("tikv address", allStores[i].GetAddress()))
-		workers.ApplyWithIDInErrorGroup(eg, func(id uint64) error {
+		defer conn.Close()
+
+		workers.ApplyOnErrorGroup(eg, func() error {
+			log.Info("read meta from tikv", zap.String("tikv address", storeAddr), zap.Uint64("store id", storeId))
 			stream, err := tikvClient.ReadRegionMeta(ectx, &recovpb.ReadRegionMetaRequest{StoreId: storeId})
 			if err != nil {
-				log.Error("read region meta failied", zap.Uint64("storeID", storeId))
+				log.Error("read region meta failed", zap.Uint64("storeID", storeId))
 				return errors.Trace(err)
 			}
 
-			tikvMeta := NewRecoveryMeta(storeId)
+			storeMeta := NewStoreMeta(storeId)
 			// for a TiKV, received the stream
 			for {
 				var meta *recovpb.RegionMeta
 				if meta, err = stream.Recv(); err == nil {
-					tikvMeta.recoveryMeta = append(tikvMeta.recoveryMeta, meta)
+					storeMeta.regionMetas = append(storeMeta.regionMetas, meta)
 				} else if err == io.EOF {
+					//read to end of stream or any unexpected EOF (e.g remote stopped), err will be catched in eg.wait.
```

Comment:
can be retried automatically?

Reply:
io.EOF is expect when server close the connection in server streaming. https://www.grpc.io/docs/languages/go/basics/.  why retry here?🤔